library/bigip_software.py

```diff
@@ -212,673 +290,1060 @@
 
-        if self.is_installed():
+    def does_checksum_match(self, md5name, filename):
+        cmd = '-c "cd /shared/images; md5sum -c /shared/images/{0}"'.format(
+            md5name
+        )
+        output = self.run_command_on_device(cmd)
+        result = str(output.commandResult).rstrip()
+        expected = "{0}: OK".format(filename)
+        if result == expected:
+            return True
+        else:
             return False
-        elif volume == self.get_active_volume():
-            raise F5ModuleError(
-                "Cannot install software or hotfixes to active volumes"
-            )
-
-        images = self.read()
-        total = sum(len(v) for v in images.itervalues())
-
-        if force:
-            if hotfix:
-                self.delete(hotfix)
-                total -= 1
-                changed = True
-
-            if software:
-                self.delete(software)
-                total -= 1
-                changed = True
-
-            if changed:
-                self.wait_for_images(total)
-                changed = False
-
-        images = self.read()
-        total = sum(len(v) for v in images.itervalues())
 
+    def download_iso_on_device(self, hotfix=False):
         if hotfix:
-            if not self.is_hotfix_available(hotfix):
-                self.upload(hotfix)
-                total += 1
-                changed = True
-
-        if not self.is_software_available(software):
-            self.upload(software)
-            total += 1
-            changed = True
-
-        if changed:
-            self.wait_for_images(total)
+            md5name = self.want.hfmd5_name
+            md5_url = self.want.hotfix_md5sum
+            name = self.name_match(md5_url, md5name, True)
+            url = self.want.hotfix
+            cmd = self.prepare_command(url, name)
+            self.run_command_on_device(cmd)
+            if not self.does_checksum_match(md5name, name):
+                err = 'ISO download failed from remote host.'
+                raise F5ModuleError(err)
+        else:
+            md5name = self.want.imgmd5_name
+            md5_url = self.want.software_md5sum
+            name = self.name_match(md5_url, md5name)
+            url = self.want.software
+            cmd = self.prepare_command(url, name)
+            self.run_command_on_device(cmd)
+            if not self.does_checksum_match(md5name, name):
+                err = 'ISO download failed from remote host.'
+                raise F5ModuleError(err)
+
+    def name_check(self, name, hotfix=False):
+        if hotfix:
+            if name != self.want.hotfix_name:
+                self.want.hotfix_name = name
+        else:
+            if name != self.want.software_name:
+                self.want.software_name = name
 
-        status = self.api.System.SoftwareManagement.get_all_software_status()
-        volumes = [x['installation_id']['install_volume'] for x in status]
+    def load_image_from_device_by_name(self):
```

Comment:
to stick with the *_on_device pattern, this should be renamed to load_image_by_name_from_device